helper/resource/testing.go

```diff
@@ -789,8 +829,41 @@ func testCheckResourceAttrSet(is *terraform.InstanceState, name string, key stri
 	return nil
 }
 
-// TestCheckResourceAttr is a TestCheckFunc which validates
-// the value in state for the given name/key combination.
+// TestCheckResourceAttr ensures a specific value is stored in state for the
+// given name and key combination. State value checking is only recommended for
+// testing Computed attributes and attribute defaults.
+//
+// For managed resources, the name parameter is combination of the resource
+// type, a period (.), and the name label. The name for the below example
+// configuration would be "myprovider_thing.example".
+//
+//     resource "myprovider_thing" "example" { ... }
+//
+// For data sources, the name parameter is a combination of the keyword "data",
+// a period (.), the data source type, a period (.), and the name label. The
+// name for the below example configuration would be
+// "data.myprovider_thing.example".
+//
+//     data "myprovider_thing" "example" { ... }
+//
+// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
+// "flatmap" syntax. Keys start with the attribute name of a top-level
+// attribute. Use the following special key syntax to inspect list, map, and
+// set attributes:
+//
+// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
+//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//              for sets.
+// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+// - .#: Number of elements in list or set
+// - .%: Number of elements in map
```

Comment:
As above, an "end to end" example would put a nice bow here.

Reply:
I've uploaded some TestCheckFunc examples.